### ISSUE TITLE
Finetuning is too slow

### ISSUE THREAD

**Issue by wisdomikezogwo** [nvidia-cosmos/cosmos-predict1#10]
Hi,

Thanks for pushing the updates. However, I wanted to know if there's any plans to use similar pre-tokenization (of video) scheme used in the previous codebase, as this codebase from my experiments is almost 9.8x slower as prior experiments over the same private dataset took ~7days on 8 A100 GPUs and now would take almost 68 days. Also I was wondering what the reason for the change was, as if there wasn't any major reasons, one (I) can modify the code to ue pretokenized latents instead of doing it on the fly as it is right now.

Thanks.

**Comment by mingyuliutw** [comment]
@wisdomikezogwo 

We did the update so that the repository can have minimal dependency. Which specific finetuning examples are you looking at to do the post-training? Autoregressive or diffusion? Let me know so that we can plan.

Typically, you could tokenize individual videos and read them. The change in the data loader should not be quite minimal.

**Comment by wisdomikezogwo** [comment]
Yeah, it was a hassle to modify the nemo code in the prior codebase as one would have to re-install any nemo modifications before training, so the change is great, nonetheless, I'm not sure why the drastic almost 10x diff in training time.

Specifics:
 - Model type: Diffusion 7B model
- I have also tried to use the pretokenized latents and the train time per sample reduced from 65s to about 54-56s which is still pretty high (now ~60 ish days instead of 68days)
- I did not change any part of the code outside the dataset class, and some parts of cosmos_predict1/diffusion/training/models/ extend_model.py, model.py, and model_image.py to not use the VAE for pretokenized inputs

happy to provide more specifics!

**Comment by SeungjunNah** [comment]
@wisdomikezogwo The diffusion text2world 7B model post-training should be 2~3 times faster now if you pull the latest version. Previously, FSDP + gradient checkpointing was used. Now, gradient checkpointing is disabled and FSDP + CP (Context Parallelism) is used instead.

**Comment by sophiahhuang** [comment]
@wisdomikezogwo We haven’t heard back from you in a while. If this is still an issue, feel free to reopen or leave a comment—happy to help further.

**Comment by wisdomikezogwo** [comment]
Apologies, made the following adjustment to [model_paralel below in video2world experiment.py](cosmos_predict1/diffusion/training/config/video2world/experiment.py) file for diffusion 7b and got the train speed back to normal

        model_parallel=dict(
            sequence_parallel=True,
            tensor_model_parallel_size=8,
            context_parallel_size=1,
        ),
